numba/tests/test_function.py

```diff
@@ -0,0 +1,575 @@
+            return i + 1
+
+        def foo(f):
+            r = f(123)
+            return r
+
+        for decor in all_func_kinds:
+            if decor in [pure_func, ctypes_func]:
+                # Skip not-yet-supported functions
+                continue
+            with self.subTest(decor=decor.__name__):
+                a_ = decor(a)
+                self.assertEqual(njit(foo)(a_), foo(a))
+
+    def test_in_call_out(self):
+
+        def a(i):
+            return i + 1
+
+        def foo(f):
+            f(123)
+            return f
+
+        for decor in all_func_kinds:
+            if decor in [pure_func, ctypes_func, njit_func, njit2_func,
+                         wap_func]:
+                # Skip not-yet-supported functions
+                continue
+            with self.subTest(decor=decor.__name__):
+                a_ = decor(a)
+                if decor is cfunc_func:
+                    self.assertEqual(njit(foo)(a_).pyfunc, foo(a))
+                else:
+                    self.assertEqual(njit(foo)(a_), foo(a))
+
+    def test_in_seq_call(self):
+
+        def a(i):
+            return i + 1
+
+        def b(i):
+            return i + 2
+
+        def foo(f, g):
+            r = 0
+            for f_ in (f, g):
+                r = r + f_(r)
+            return r
+
+        for decor in all_func_kinds:
+            if decor in [pure_func, ctypes_func, njit_func, njit2_func]:
+                # Skip not-yet-supported functions
+                continue
+            with self.subTest(decor=decor.__name__):
+                a_ = decor(a)
+                b_ = decor(b)
+                self.assertEqual(njit(foo)(a_, b_), foo(a, b))
+
+    def test_in_ns_seq_call(self):
+
+        def a(i):
+            return i + 1
+
+        def b(i):
+            return i + 2
+
+        def mkfoo(b_):
+            def foo(f):
+                r = 0
+                for f_ in (f, b_):
+                    r = r + f_(r)
+                return r
+            return foo
+
+        for decor in all_func_kinds:
+            if decor in [pure_func, ctypes_func, njit_func, njit2_func,
+                         wap_func]:
+                # Skip not-yet-supported functions
+                continue
+            with self.subTest(decor=decor.__name__):
+                a_ = decor(a)
+                b_ = decor(b)
+                self.assertEqual(njit(mkfoo(b_))(a_), mkfoo(b)(a))
+
+    def test_ns_call(self):
+
+        def a(i):
+            return i + 1
+
+        def mkfoo(a_):
+            def foo():
+                r = a_(123)
+                return r
+            return foo
+
+        for decor in all_func_kinds:
+            if decor in [pure_func, ctypes_func, wap_func]:
+                # Skip not-yet-supported functions
+                continue
+            with self.subTest(decor=decor.__name__):
+                a_ = decor(a)
+                self.assertEqual(njit(mkfoo(a_))(), mkfoo(a)())
+
+    def test_ns_out(self):
```

Comment:
I feel that this test name needs expanding, or a doc string for what it is testing.

Reply:
Added doc string.